README
Data-Babel

This module translates biological identifiers based on information
contained in a database.  The principal method is 'translate' which
converts identifiers of one type into identifiers of one or more
output types.  Each Data::Babel object provides a unique mapping over
a set of identifier types. The system as a whole can contain multiple
Data::Babel objects; these may share some or all identifier types, and
may provide the same or different mappings over the shared types.

INSTALLATION

To install this module, run the following commands:

	perl Build.PL
	./Build
	./Build test
	./Build install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Data::Babel

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Data-Babel

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Data-Babel

    CPAN Ratings
        http://cpanratings.perl.org/d/Data-Babel

    Search CPAN
        http://search.cpan.org/dist/Data-Babel/


LICENSE AND COPYRIGHT

Copyright (C) 2010 Institute for Systems Biology

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for Data-Babel

1.13_07 2013-10-13
	Diagnosed FAILs seen by David Cantrell. Caused by MySQL bug in duplicate
	  removal (SELECT DISTINCT) with repeated output columns. Bug present in
	  MySQL 5.0.32, fixed in or before 5.0.86
	Test changes
	  - Added test for MySQL bug in duplicate removal (SELECT DISTINCT) with 
	    repeated output column in 000.reqs
	  - Removed debugging code for FAILs seen by David Cantrell

1.13_06 2013-10-05
	Still working on David's FAILs

1.13_05 2013-09-29
	New features
	- More complex filters for translate and related methods
	Updated depdendencies
        Bug fixes
        - maptables attribute of recycled idtype not reset
	Still trying to track down FAILs seen by David Cantrell
	  (reports 34101829, 34102877, 34457224)
	Test changes
	- More test cleanup subsequent to fixing of 'implicit masters' bug
        - Fixed boilerplate & pod tests


1.13_04 2013-09-23
        Bug fixes
        - Implicit master created in application and passed into Babel
	  constructor not processed by make_implicit_masters and thus not set up
	  properly
	Still trying to track down FAILs seen by David Cantrell
	  (reports 34101829, 34102877, 34457224)
	Also includes code and tests for Filter object and code stub for
	  testing translate with Filter object, but translate not yet ported to
	  Filter object

1.13_03 2013-09-17
	Another debugging release to track down FAILs seen by David Cantrell
	  (reports 34101829, 34102877, 34457224)
	Also includes documentation for new translate filters

1.13_02 2013-09-15
	This is a debugging release to track down FAILs seen by David Cantrell
	  (reports 34101829, 34102877)	

1.13_01 2013-09-02
	New features
	- 'check_contents' method - checks whether explict Masters contain all
	  ids in the underlying MapTables
        Bug fixes
        - 'apostrophe' bug: ID containing apostrophe triggered bug in partial
	  duplicate removal
	Test changes
	- Added test for schema mechanics
	Documentation changes
	- Deprecated obsolete workflow-related attributes and removed them from
	  examples
	- Deprecated obsolete maptable header macro - it was just a bad idea...
----------
1.12    2013-07-31
	Bug fixes
	- Fixed partial duplicates bug in which translate returned rows that 
	  were identical on all non-null columns, differing only in that one row
	  had nulls where the other had a real value
	Test changes
	- Cleaned up tests to improve CPAN install speed
	- Changed the scheme for telling CPAN that it's futile to run the tests
	  because MySQL is not accessible. I now check key requirements in a 
	  test script and skip out of the test suite if the requirements are not
	  met.  The main requirement that I check is that MySQL be accessible.
	  The test suite reports PASS in this case!! It would be far better to
	  eport NA or UNKNOWN, but I don't know how to do this robustly.
----------
1.11    2013-06-18
	New features
	- can specify history in IdType, rather than just in Master
	- new 'show_schema_graph' method
	Bug fixes
	- in 'validate', check for unmatched ids has to be case insensitive 
	  because MySQL does case intensive comparisons, e.g., searching for 
	  'htt' as gene_symbol
    	- detect isolated IdTypes, ie, IdTypes not contained in any MapTable.
          previously, load_implicit_masters would generate illegal empty query
----------
1.10	2012-12-02
	New features
	* new 'translate' options
	  - filters
	  - count
	  - validate
	* new methods
	  - load_implicit_masters
	  - count
	  - validate
	* histories
	    Master tables can hold old and current values for ids. translate,
	    count, validate use old values if they exist
	Bug fixes
	- 'translate' with empty input_ids. was generating illegal SQL, now
	  returns empty table
	- examples/babel.pl used deprecated qw(...) as parentheses. should have
	  been fixed in 1.02.  sorry	
	- 'translate' with empty input_ids would sometimes return NULL id
	- lingering bug in 098.example. test was calling example script with
	  wrong Perl interpreter and without -Mblib. thanks to CPAN smoke 
	  testers for finding this bug!!
----------
1.03    2012-08-21
	Added test for version number in POD
	Removed redundant 'dist_version_from' param from Build.PL

1.02    2012-08-02
        Changed tests to correct deprecated use of qw(...) as parentheses

1.01    2011-01-22
        Added 'translate all' (more precisely, the input_ids_all option to
	  translate) to generate a complete mapping of the input type to the
          output types.

1.00    2010-11-29
        First CPAN version, released on an unsuspecting world.

1.00_03 2010-11-19
	Added Module::Build to build_requires
	Added view operations to chk_mysql in Build.PL on the theory that FAILs 
	  are due to inability to create views

1.00_02 2010-11-16
	Added Class::AutoDB to requires

0.90    2010-08-22
        First version, released for internal use.



lib/Data/Babel/IdType.pm
package Data::Babel::IdType;
#################################################################################
#
# Author:  Nat Goodman
# Created: 10-07-26
# $Id: 
#
# Copyright 2010 Institute for Systems Biology
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of either: the GNU General Public License as published
# by the Free Software Foundation; or the Artistic License.
#
# See http://dev.perl.org/licenses/ for more information.
#
#################################################################################
use strict;
use Carp;
use Class::AutoClass;
use vars qw(@AUTO_ATTRIBUTES @OTHER_ATTRIBUTES @CLASS_ATTRIBUTES %SYNONYMS %DEFAULTS %AUTODB);
use base qw(Data::Babel::Base);

# babel, name, id, autodb, verbose - methods defined in Base
@AUTO_ATTRIBUTES=qw(master referent defdb meta format sql_type internal _maptables);
@OTHER_ATTRIBUTES=qw(display_name maptables external tablename history);
@CLASS_ATTRIBUTES=qw();
%SYNONYMS=(perl_format=>'format');
%DEFAULTS=(_maptables=>{});
%AUTODB=
  (-collection=>'IdType',
   -keys=>qq(name string,display_name string,referent string,defdb string,meta string,
             perl_format string,sql_type string,internal string));
   
Class::AutoClass::declare;

# must run after Babel initialized
sub connect_master {
  my $self=shift;
  my $master_name=$self->name.'_master'; # append '_master' to idtype name
  $self->{master}=$self->babel->name2master($master_name)
    or confess 'Trying to connect IdType '.$self->name.' to non-existent Master';
  # NG 13-06-11: propogate history if it exists
  if (exists $self->{history}) {
    $self->master->history($self->{history}) unless $self->master->history;
    delete $self->{history};
  }
}
# must run after Babel initialized
# NG 13-09-23: when re-using IdType in new Babel, duplicates can arise
sub add_maptable {
  my($self,$maptable)=@_;
  $self->_maptables->{$maptable->name}=$maptable;
}
# NG 13-09-23: when re-using IdType in new Babel, duplicate maptables can arise
sub maptables {
  my $self=shift;
  if (@_) {
    my %maptables=map {$_->name=>$_} @{$_[0]};
    $self->{_maptables}=\%maptables;
  }
  [values %{$self->{_maptables}}];
}
# degree is number of MapTables containing this IdType
sub degree {scalar @{shift->maptables}}

our $WARN_INTERNAL=": FOR INTERNAL USE ONLY";
sub display_name {
  my $self=shift;
  my $display_name=@_? $self->{_display_name}=shift: $self->{_display_name};
  $self->internal? "$display_name$WARN_INTERNAL": $display_name;
}

# opposite of internal
sub external {
  my $self=shift;
  @_? !$self->internal(!$_[0]): !$self->internal;
}
# tablename - delegates to master
sub tablename {
  my $self=shift;
  defined $self->master && $self->master->tablename(@_);
}
# history - delegate to master
# NG 13-06-11: maintain history here until connected to master
sub history {
  my $self=shift;
  return $self->master->history(@_) if defined $self->master;
  @_? $self->{history}=$_[0]: $self->{history};
}

# NG 10-08-08. sigh.'verbose' in Class::AutoClass::Root conflicts with method in Base
#              because AutoDB splices itself onto front of @ISA.
sub verbose {Data::Babel::Base::verbose(@_)}
1;
